src/kontrol/prove.py

```diff
@@ -241,105 +246,132 @@ def init_and_run_proof(_init_problem: tuple[str, str, int]) -> tuple[bool, list[
-                init_node_id,
-                target_node_id,
-                {},
-                bmc_depth,
-                proof_dir=save_directory,
-            )
-        else:
-            apr_proof = APRProof(test_id, kcfg, [], init_node_id, target_node_id, {}, proof_dir=save_directory)
+    if bmc_depth is not None:
+        apr_proof = APRBMCProof(
+            test.id,
+            kcfg,
+            [],
+            init_node_id,
+            target_node_id,
+            {},
+            bmc_depth,
+            proof_dir=foundry.proofs_dir,
+        )
+    else:
+        apr_proof = APRProof(test.id, kcfg, [], init_node_id, target_node_id, {}, proof_dir=foundry.proofs_dir)
 
     apr_proof.write_proof_data()
     return apr_proof
 
 
+def _load_setup_cterm(foundry: Foundry, contract: Contract) -> CTerm:
+    latest_version = foundry.latest_proof_version(f'{contract.name}.setUp()')
+    setup_digest = f'{contract.name}.setUp():{latest_version}'
+    apr_proof = APRProof.read_proof_data(foundry.proofs_dir, setup_digest)
+    target = apr_proof.kcfg.node(apr_proof.target)
+    target_states = apr_proof.kcfg.covers(target_id=target.id)
+    if len(target_states) == 0:
+        raise ValueError(
+            f'setUp() function for {apr_proof.id} did not reach the end of execution. Maybe --max-iterations is too low?'
+        )
+    if len(target_states) > 1:
+        raise ValueError(f'setUp() function for {apr_proof.id} branched and has {len(target_states)} target states.')
+    cterm = single(target_states).source.cterm
+    return cterm
+
+
+def method_to_initialized_cfg(
+    foundry: Foundry,
+    test: FoundryTest,
+    kcfg_explore: KCFGExplore,
+    *,
+    setup_cterm: CTerm | None = None,
+    simplify_init: bool = True,
+) -> tuple[KCFG, int, int]:
+    _LOGGER.info(f'Initializing KCFG for test: {test.id}')
+
+    empty_config = foundry.kevm.definition.empty_config(GENERATED_TOP_CELL)
+    kcfg, init_node_id, target_node_id = _method_to_cfg(
+        empty_config,
+        test.contract,
+        test.method,
+        setup_cterm,
+    )
+
+    _LOGGER.info(f'Expanding macros in initial state for test: {test.id}')
+    init_term = kcfg.node(init_node_id).cterm.kast
+    init_term = KDefinition__expand_macros(foundry.kevm.definition, init_term)
+    init_cterm = CTerm.from_kast(init_term)
+    _LOGGER.info(f'Computing definedness constraint for test: {test.id}')
+    init_cterm = kcfg_explore.cterm_assume_defined(init_cterm)
+    kcfg.replace_node(init_node_id, init_cterm)
+
+    _LOGGER.info(f'Expanding macros in target state for test: {test.id}')
+    target_term = kcfg.node(target_node_id).cterm.kast
+    target_term = KDefinition__expand_macros(foundry.kevm.definition, target_term)
+    target_cterm = CTerm.from_kast(target_term)
+    kcfg.replace_node(target_node_id, target_cterm)
+
+    if simplify_init:
+        _LOGGER.info(f'Simplifying KCFG for test: {test.id}')
+        kcfg_explore.simplify(kcfg, {})
```

Comment:
We should always do this actually, and remove this option.

Reply:
Opened an issue: https://github.com/runtimeverification/kontrol/issues/67